freqtrade/freqai/data_drawer.py

```diff
@@ -0,0 +1,589 @@
+            df[f"{label}_std"].iloc[-1] = dk.data["labels_std"][label]
+        # df['prediction'].iloc[-1] = predictions[-1]
+        df["do_predict"].iloc[-1] = do_preds[-1]
+
+        if self.freqai_info.get("feature_parameters", {}).get("DI_threshold", 0) > 0:
+            df["DI_values"].iloc[-1] = dk.DI_values[-1]
+
+        # append the new predictions to persistent storage
+        if pair in self.historic_predictions:
+            self.historic_predictions[pair].iloc[-1] = df[label].iloc[-1]
+
+        if length_difference < 0:
+            prepend_df = pd.DataFrame(
+                np.zeros((abs(length_difference) - 1, len(df.columns))), columns=df.columns
+            )
+            df = pd.concat([prepend_df, df], axis=0)
+
+    def attach_return_values_to_return_dataframe(
+            self, pair: str, dataframe: DataFrame) -> DataFrame:
+        """
+        Attach the return values to the strat dataframe
+        :param dataframe: DataFrame = strategy dataframe
+        :return: DataFrame = strat dataframe with return values attached
+        """
+        df = self.model_return_values[pair]
+        to_keep = [col for col in dataframe.columns if not col.startswith("&")]
+        dataframe = pd.concat([dataframe[to_keep], df], axis=1)
+        return dataframe
+
+    def return_null_values_to_strategy(self, dataframe: DataFrame, dk) -> None:
+        """
+        Build 0 filled dataframe to return to strategy
+        """
+
+        dk.find_features(dataframe)
+
+        for label in dk.label_list:
+            dataframe[label] = 0
+            dataframe[f"{label}_mean"] = 0
+            dataframe[f"{label}_std"] = 0
+
+        # dataframe['prediction'] = 0
+        dataframe["do_predict"] = 0
+
+        if self.freqai_info.get("feature_parameters", {}).get("DI_threshold", 0) > 0:
+            dataframe["DI_value"] = 0
+
+        dk.return_dataframe = dataframe
+
+    def purge_old_models(self) -> None:
+
+        model_folders = [x for x in self.full_path.iterdir() if x.is_dir()]
+
+        pattern = re.compile(r"sub-train-(\w+)_(\d{10})")
+
+        delete_dict: Dict[str, Any] = {}
+
+        for dir in model_folders:
+            result = pattern.match(str(dir.name))
+            if result is None:
+                break
+            coin = result.group(1)
+            timestamp = result.group(2)
+
+            if coin not in delete_dict:
+                delete_dict[coin] = {}
+                delete_dict[coin]["num_folders"] = 1
+                delete_dict[coin]["timestamps"] = {int(timestamp): dir}
+            else:
+                delete_dict[coin]["num_folders"] += 1
+                delete_dict[coin]["timestamps"][int(timestamp)] = dir
+
+        for coin in delete_dict:
+            if delete_dict[coin]["num_folders"] > 2:
+                sorted_dict = collections.OrderedDict(
+                    sorted(delete_dict[coin]["timestamps"].items())
+                )
+                num_delete = len(sorted_dict) - 2
```

Comment:
wouldn't a time-based approach be better / more reliable?
Currently i think this works by removing all files (sorted by timestamp) except for the last 2 - but in reality, you'll know old a model is allowed to be.